# Data Wrangling: Exercises

---
Erasmus Q-Intelligence B.V.

---

Below is a set of exercises. For every question, an empty code cell is provided where you can add your code that serves as an answer to the question. If there are additional questions regarding interpreting what you see, you can do so by adding comments (start with `#`), or adding a *Markdown* cell below or above the *Code* cell, and providing your findings there.

## Exercise 1: query()

Use the flights data from the file flights.csv. The data set contains information on flights
from airports in New York in 2013.

### a
Find all flights without any delay on departure (dep delay) and with at least 2 hours delay
on arrival (arr delay)

In [1]:
pip install --upgrade setuptools

   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.2/1.2 MB 9.0 MB/s  0:00:00
Note: you may need to restart the kernel to use updated packages.


In [3]:
%pip install nycflights13

from nycflights13 import airlines
airlines


     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 8.7/8.7 MB 10.2 MB/s  0:00:00 eta 0:00:01
  Installing build dependencies ... done
  Getting requirements to build wheel ... done
  Preparing metadata (pyproject.toml) ... done
  Created wheel for nycflights13: filename=nycflights13-0.0.3-py3-none-any.whl size=8732783 sha256=a910e7664149f5be1d298f90f656d6f2ba468fd66ee7a6449595f2779bb1daee
  Stored in directory: /Users/adilvural/Library/Caches/pip/wheels/64/33/b8/55dbea9d199e6f4869488e0b029d09d91c34c4f4706f232253
Successfully built nycflights13
Note: you may need to restart the kernel to use updated packages.


/Library/Frameworks/Python.framework/Versions/3.13/lib/python3.13/site-packages/nycflights13/__init__.py:2: UserWarning: pkg_resources is deprecated as an API. See https://setuptools.pypa.io/en/latest/pkg_resources.html. The pkg_resources package is slated for removal as early as 2025-11-30. Refrain from using this package or pin to Setuptools<81.
  from pkg_resources import resource_filename as _rf


carrier                         name
0       9E            Endeavor Air Inc.
1       AA       American Airlines Inc.
2       AS         Alaska Airlines Inc.
3       B6              JetBlue Airways
4       DL         Delta Air Lines Inc.
5       EV     ExpressJet Airlines Inc.
6       F9       Frontier Airlines Inc.
7       FL  AirTran Airways Corporation
8       HA       Hawaiian Airlines Inc.
9       MQ                    Envoy Air
10      OO        SkyWest Airlines Inc.
11      UA        United Air Lines Inc.
12      US              US Airways Inc.
13      VX               Virgin America
14      WN       Southwest Airlines Co.
15      YV           Mesa Airlines Inc.

### b
Find all flights of carriers United (UA), American (AA) and Delta (DL)

In [9]:
from nycflights13 import flights

carriers = ['UA', 'AA', 'DL']
flights_filtered = flights[flights['carrier'].isin(carriers)]
flights_filtered


year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
4       2013      1    1     554.0             600       -6.0     812.0   
5       2013      1    1     554.0             558       -4.0     740.0   
...      ...    ...  ...       ...             ...        ...       ...   
336737  2013      9   30    2105.0            2106       -1.0    2329.0   
336744  2013      9   30    2121.0            2100       21.0    2349.0   
336751  2013      9   30    2140.0            2140        0.0      10.0   
336755  2013      9   30    2149.0            2156       -7.0    2245.0   
336762  2013      9   30    2233.0            2113       80.0     112.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
5                  728       12.0      UA    1696  N39463    EWR  ORD   
...                ...        ...     ...     ...     ...    ...  ...   
336737            2354      -25.0      UA     475  N477UA    EWR  IAH   
336744              14      -25.0      DL    2363  N193DN    JFK  LAX   
336751              40      -30.0      AA     185  N335AA    JFK  LAX   
336755            2308      -23.0      UA     523  N813UA    EWR  BOS   
336762              30       42.0      UA     471  N578UA    EWR  SFO   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
5          150.0       719     5      58  2013-01-01T10:00:00Z  
...          ...       ...   ...     ...                   ...  
336737     175.0      1400    21       6  2013-10-01T01:00:00Z  
336744     296.0      2475    21       0  2013-10-01T01:00:00Z  
336751     298.0      2475    21      40  2013-10-01T01:00:00Z  
336755      37.0       200    21      56  2013-10-01T01:00:00Z  
336762     318.0      2565    21      13  2013-10-01T01:00:00Z  

[139504 rows x 19 columns]

In [10]:
flights_filtered = flights.query("carrier in ['UA', 'AA', 'DL']")
flights_filtered

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
4       2013      1    1     554.0             600       -6.0     812.0   
5       2013      1    1     554.0             558       -4.0     740.0   
...      ...    ...  ...       ...             ...        ...       ...   
336737  2013      9   30    2105.0            2106       -1.0    2329.0   
336744  2013      9   30    2121.0            2100       21.0    2349.0   
336751  2013      9   30    2140.0            2140        0.0      10.0   
336755  2013      9   30    2149.0            2156       -7.0    2245.0   
336762  2013      9   30    2233.0            2113       80.0     112.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
5                  728       12.0      UA    1696  N39463    EWR  ORD   
...                ...        ...     ...     ...     ...    ...  ...   
336737            2354      -25.0      UA     475  N477UA    EWR  IAH   
336744              14      -25.0      DL    2363  N193DN    JFK  LAX   
336751              40      -30.0      AA     185  N335AA    JFK  LAX   
336755            2308      -23.0      UA     523  N813UA    EWR  BOS   
336762              30       42.0      UA     471  N578UA    EWR  SFO   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
5          150.0       719     5      58  2013-01-01T10:00:00Z  
...          ...       ...   ...     ...                   ...  
336737     175.0      1400    21       6  2013-10-01T01:00:00Z  
336744     296.0      2475    21       0  2013-10-01T01:00:00Z  
336751     298.0      2475    21      40  2013-10-01T01:00:00Z  
336755      37.0       200    21      56  2013-10-01T01:00:00Z  
336762     318.0      2565    21      13  2013-10-01T01:00:00Z  

[139504 rows x 19 columns]

In [12]:
# Definieer de lijst van carriers
carriers = ['UA', 'AA', 'DL']
# Gebruik de query-methode met de @-notatie om de lijst te verwijzen
flights_filtered = flights.query("carrier in @carriers")
flights_filtered

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
1       2013      1    1     533.0             529        4.0     850.0   
2       2013      1    1     542.0             540        2.0     923.0   
4       2013      1    1     554.0             600       -6.0     812.0   
5       2013      1    1     554.0             558       -4.0     740.0   
...      ...    ...  ...       ...             ...        ...       ...   
336737  2013      9   30    2105.0            2106       -1.0    2329.0   
336744  2013      9   30    2121.0            2100       21.0    2349.0   
336751  2013      9   30    2140.0            2140        0.0      10.0   
336755  2013      9   30    2149.0            2156       -7.0    2245.0   
336762  2013      9   30    2233.0            2113       80.0     112.0   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
1                  830       20.0      UA    1714  N24211    LGA  IAH   
2                  850       33.0      AA    1141  N619AA    JFK  MIA   
4                  837      -25.0      DL     461  N668DN    LGA  ATL   
5                  728       12.0      UA    1696  N39463    EWR  ORD   
...                ...        ...     ...     ...     ...    ...  ...   
336737            2354      -25.0      UA     475  N477UA    EWR  IAH   
336744              14      -25.0      DL    2363  N193DN    JFK  LAX   
336751              40      -30.0      AA     185  N335AA    JFK  LAX   
336755            2308      -23.0      UA     523  N813UA    EWR  BOS   
336762              30       42.0      UA     471  N578UA    EWR  SFO   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
1          227.0      1416     5      29  2013-01-01T10:00:00Z  
2          160.0      1089     5      40  2013-01-01T10:00:00Z  
4          116.0       762     6       0  2013-01-01T11:00:00Z  
5          150.0       719     5      58  2013-01-01T10:00:00Z  
...          ...       ...   ...     ...                   ...  
336737     175.0      1400    21       6  2013-10-01T01:00:00Z  
336744     296.0      2475    21       0  2013-10-01T01:00:00Z  
336751     298.0      2475    21      40  2013-10-01T01:00:00Z  
336755      37.0       200    21      56  2013-10-01T01:00:00Z  
336762     318.0      2565    21      13  2013-10-01T01:00:00Z  

[139504 rows x 19 columns]

## Exercise 2: sort_values()


### a
Sort the flights dataset in such a way that flights with the highest delay at arrival
(arr delay) come first.

In [14]:
# Assuming 'flights' is your DataFrame and 'arr_delay' is the relevant column
flights_sorted = flights.sort_values(by='arr_delay', ascending=False)
flights_sorted

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
7072    2013      1    9     641.0             900     1301.0    1242.0   
235778  2013      6   15    1432.0            1935     1137.0    1607.0   
8239    2013      1   10    1121.0            1635     1126.0    1239.0   
327043  2013      9   20    1139.0            1845     1014.0    1457.0   
270376  2013      7   22     845.0            1600     1005.0    1044.0   
...      ...    ...  ...       ...             ...        ...       ...   
336771  2013      9   30       NaN            1455        NaN       NaN   
336772  2013      9   30       NaN            2200        NaN       NaN   
336773  2013      9   30       NaN            1210        NaN       NaN   
336774  2013      9   30       NaN            1159        NaN       NaN   
336775  2013      9   30       NaN             840        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
7072              1530     1272.0      HA      51  N384HA    JFK  HNL   
235778            2120     1127.0      MQ    3535  N504MQ    JFK  CMH   
8239              1810     1109.0      MQ    3695  N517MQ    EWR  ORD   
327043            2210     1007.0      AA     177  N338AA    JFK  SFO   
270376            1815      989.0      MQ    3075  N665MQ    JFK  CVG   
...                ...        ...     ...     ...     ...    ...  ...   
336771            1634        NaN      9E    3393     NaN    JFK  DCA   
336772            2312        NaN      9E    3525     NaN    LGA  SYR   
336773            1330        NaN      MQ    3461  N535MQ    LGA  BNA   
336774            1344        NaN      MQ    3572  N511MQ    LGA  CLE   
336775            1020        NaN      MQ    3531  N839MQ    LGA  RDU   

        air_time  distance  hour  minute             time_hour  
7072       640.0      4983     9       0  2013-01-09T14:00:00Z  
235778      74.0       483    19      35  2013-06-15T23:00:00Z  
8239       111.0       719    16      35  2013-01-10T21:00:00Z  
327043     354.0      2586    18      45  2013-09-20T22:00:00Z  
270376      96.0       589    16       0  2013-07-22T20:00:00Z  
...          ...       ...   ...     ...                   ...  
336771       NaN       213    14      55  2013-09-30T18:00:00Z  
336772       NaN       198    22       0  2013-10-01T02:00:00Z  
336773       NaN       764    12      10  2013-09-30T16:00:00Z  
336774       NaN       419    11      59  2013-09-30T15:00:00Z  
336775       NaN       431     8      40  2013-09-30T12:00:00Z  

[336776 rows x 19 columns]

### b
Sort the flights dataset in such a way that flights with tailnum equal to NaN come first.

In [16]:
# Assuming 'flights' is your DataFrame
flights_sorted = flights.sort_values(by='tailnum', key=lambda x: x.isna(), ascending=True)
flights_sorted

year  month  day  dep_time  sched_dep_time  dep_delay  arr_time  \
0       2013      1    1     517.0             515        2.0     830.0   
224090  2013      6    3     856.0             900       -4.0    1130.0   
224089  2013      6    3     855.0             834       21.0    1123.0   
224088  2013      6    3     855.0             840       15.0    1112.0   
224087  2013      6    3     855.0             847        8.0    1220.0   
...      ...    ...  ...       ...             ...        ...       ...   
120740  2013      2   11       NaN            2005        NaN       NaN   
247667  2013      6   27       NaN             820        NaN       NaN   
247668  2013      6   27       NaN            2000        NaN       NaN   
14929   2013      1   17       NaN             645        NaN       NaN   
248629  2013      6   28       NaN            1700        NaN       NaN   

        sched_arr_time  arr_delay carrier  flight tailnum origin dest  \
0                  819       11.0      UA    1545  N14228    EWR  IAH   
224090            1126        4.0      UA    1643  N33103    EWR  DEN   
224089            1018       65.0      EV    4344  N14953    EWR  MEM   
224088            1038       34.0      EV    4297  N12157    EWR  DTW   
224087            1158       22.0      UA    1401  N37274    EWR  MIA   
...                ...        ...     ...     ...     ...    ...  ...   
120740            2258        NaN      UA     647     NaN    EWR  MCO   
247667            1042        NaN      9E    4248     NaN    JFK  CHS   
247668            2127        NaN      9E    4287     NaN    JFK  BWI   
14929              952        NaN      UA     719     NaN    EWR  DFW   
248629            1923        NaN      9E    3398     NaN    JFK  CVG   

        air_time  distance  hour  minute             time_hour  
0          227.0      1400     5      15  2013-01-01T10:00:00Z  
224090     216.0      1605     9       0  2013-06-03T13:00:00Z  
224089     147.0       946     8      34  2013-06-03T12:00:00Z  
224088      87.0       488     8      40  2013-06-03T12:00:00Z  
224087     151.0      1085     8      47  2013-06-03T12:00:00Z  
...          ...       ...   ...     ...                   ...  
120740       NaN       937    20       5  2013-02-12T01:00:00Z  
247667       NaN       636     8      20  2013-06-27T12:00:00Z  
247668       NaN       184    20       0  2013-06-28T00:00:00Z  
14929        NaN      1372     6      45  2013-01-17T11:00:00Z  
248629       NaN       589    17       0  2013-06-28T21:00:00Z  

[336776 rows x 19 columns]

## Exercise 3: filter and loc()


### a
 Select all columns that end with time or with delay.

In [19]:
selected_columns = flights.columns[flights.columns.str.endswith(('time', 'delay'))]
selected_columns

Index(['dep_time', 'sched_dep_time', 'dep_delay', 'arr_time', 'sched_arr_time',
       'arr_delay', 'air_time'],
      dtype='object')

In [18]:
# Assuming 'flights' is your DataFrame
selected_columns = flights.filter(regex='(time|delay)$')
selected_columns


dep_time  sched_dep_time  dep_delay  arr_time  sched_arr_time  \
0          517.0             515        2.0     830.0             819   
1          533.0             529        4.0     850.0             830   
2          542.0             540        2.0     923.0             850   
3          544.0             545       -1.0    1004.0            1022   
4          554.0             600       -6.0     812.0             837   
...          ...             ...        ...       ...             ...   
336771       NaN            1455        NaN       NaN            1634   
336772       NaN            2200        NaN       NaN            2312   
336773       NaN            1210        NaN       NaN            1330   
336774       NaN            1159        NaN       NaN            1344   
336775       NaN             840        NaN       NaN            1020   

        arr_delay  air_time  
0            11.0     227.0  
1            20.0     227.0  
2            33.0     160.0  
3           -18.0     183.0  
4           -25.0     116.0  
...           ...       ...  
336771        NaN       NaN  
336772        NaN       NaN  
336773        NaN       NaN  
336774        NaN       NaN  
336775        NaN       NaN  

[336776 rows x 7 columns]

## Exercise 4: assign()
### a

The sched_dep_time column is formatted like a digital clock, where 515 actually means 5:15. Use
assign() to create a column sched_dep_time_in_mins which gives the total number of minutes since
midnight.

In [20]:
# Assuming 'flights' is your DataFrame
flights = flights.assign(
    sched_dep_time_in_mins=lambda x: (x['sched_dep_time'] // 100) * 60 + (x['sched_dep_time'] % 100)
)

## Exercise 5. agg() and groupby()

### a
Calculate the number of canceled flights by month, the total number of flights per month and
the amount of canceled flights per month. Sort the new data.frame such that months
with the highest amount of canceled flights come first. Canceled flights either have no
dep time or do not have arr time.

Use: flights_with_status = flights.assign(is_canceled = (flights['dep_time'].isna() | flights['arr_time'].isna())) to create the column that gets the canceled flights

In [ ]:
flights_with_status = flights.assign(is_canceled = (flights['dep_time'].isna() | flights['arr_time'].isna()))



### b
Calculate the minimum, maximum, mean, median and standard deviation of air_time by
dest. Sort the new data.frame such that destinations with the highest mean air time
comes first.

## Exercise 6. groupby(), query() and assign()

### a
Add a column dep_delay_lag to flights, with dep delay of the previous flight from that
airport. (We use the .shift function here, which isn't covered in the lessons)

Show the dataframe with the origin, dest, dep_delay and dep_delay_lag columns

In [ ]:
# 1. Shift values down by 1 position within each group (airport) 2.  Creates a "previous flight" reference for each flight from the same origin airport
# 3. The first flight from each airport gets NaN since there's no previous flight
flightsdepdelay = flights.assign(dep_delay_lag = flights.groupby('origin')['dep_delay'].shift(1))



### b 
Filter the observations such that you only have those where neither dep delay, nor dep delay lag is missing. Use the dataframe: flightsdepdelay from 6.a

Tip use .notnull()

### c

Find the average of dep_delay, separate for each value of dep_delay_lag and for each origin
using groupby. Name this variable dep delay avg. Use the dataframe from 6.b

### d 
Visualize the relation (using a scatter plot) between dep delay lag and dep delay avg, where
the color of the point depends on origin.

In [ ]:
import seaborn as sns



## Exercise 7. combined
Consider the forsale.Rds-dataset from Canvas. It contains information about houses for sale
in Rotterdam area during the recent past.

### a 
Load the data from canvas

In [ ]:
forsale = pd.read_csv("forsale.csv")
forsale

### b 
Show the first 10 observations, with only the variables postcode, city, suburb,
asking price and living area.

### c 
Add a variable to the original dataframe forsale that contains the price per square metre.
Call this column price_m2.

### d 
Calculate the number of houses in this dataset in each city, as well as the minimum, average
and maximum asking price. Which city is the cheapest, on average?

### e 
Create a subset of the forsale dataset called forsalesubset, which contains only houses that satisfy
the following criteria: it is within the city of Rotterdam, it has at least 3 bedrooms and 2
bathrooms and costs no more than 400.000 euro.

### g 
Sort the new dataframe such that those with the highest asking_price occur first. 

## Exercise 8. nunique & isna

### a 
Load the penguins-data from penguins.csv. The data is on penguin
species in Palmer Archipelago.

In [ ]:
penguins = pd.read_csv("penguins.csv")
penguins

### b
 Get the number of distinct categories for all variables, using the function nunique().

### c 
Get the number of distinct categories, only for species, island, year and sex.

### d
Get the number of missing values, only for the variables of type factor (categorical variables). Use isna()

### e 
Get the number of missing values for all columns that start with bill_

## Exercise 9. join
### a 
Add name from the airlines data set (available from the nycflights13-
package, you have to pip install nycflights13 if you haven't already) to the flights-dataframe. Which
key should be used?

In [ ]:
from nycflights13 import airlines
airlines

## Exercise 10. tidy data - Melt
### a 
Load the data melt_example.csv from Canvas. The data is on the distribution of male and female in the US navy.

In [ ]:
melt = pd.read_csv("melt_example.csv")
melt = melt.filter(melt.columns[1:]) # Removes the first column which is a fake index leftover from R
melt

 Take all the columns (except paygrade) and "melt" them into two new columns - one for the column names (gender_marital) and one for their values (amount).

## Exercise 11. tidy data - Pivot
### a 
Load the data on NBA players and points scored (fictive) from Canvas.

In [ ]:
nba = pd.read_csv("nba.csv")
nba = nba.filter(nba.columns[1:]) # Removes the first column which is a fake index leftover from R
nba

### b
Use pivot to transform the long format data into wide format, where each unique value in the 'day' column becomes a separate column, with 'player' and 'team' as row identifiers and 'points' as the values."